File: 프로그래머스 고득점 kit/동적계획법/N으로 표현.py
# bfs로 모두 완탐하는데 틀리네;;
# 이건 다이나믹으로 풀어야하는데 모르겠누...
# 다른 사람도 틀림;;
from collections import deque


def solution(N, number):
    q = deque([(N, 1)])

    minimum = -1

    while q:
        curr, cnt = q.popleft()

        if curr == number and minimum == -1:
            minimum = cnt

        if cnt > 8:
            continue

        q.append((int(str(curr) + str(N)), cnt+1))
        q.append((curr + N, cnt+1))
        q.append((curr - N, cnt+1))
        q.append((curr * N, cnt+1))
        q.append((curr // N, cnt+1))

    return minimum


print(solution(5, 12) == 4)
print(solution(2, 11) == 3)

File: 프로그래머스 고득점 kit/동적계획법/N으로 표현.js
/** @param {number} N @param {number} number */
function solution(N, number) {
  let ministCnt = Number.MAX_SAFE_INTEGER

  /** @param {number} curr @param {number} cnt */
  function dfs(curr, cnt) {
    if (curr === number) ministCnt = Math.min(cnt, ministCnt)
    if (cnt === 8) return
    console.log(curr)

    dfs(Number(String(curr) + String(N)), cnt + 1)
    dfs(curr + N, cnt + 1)
    dfs(curr - N, cnt + 1)
    dfs(curr * N, cnt + 1)
    dfs(Math.floor(curr / N), cnt + 1)
  }

  dfs(N, 1)

  return ministCnt > 8 ? -1 : ministCnt
}

console.log(solution(5, 12) === 4)
console.log(solution(2, 11) === 3)
